ctapipe/image/extractor.py

```diff
@@ -1681,7 +1681,9 @@ def __call__(
             d_waveforms = deconvolve(waveforms, 0.0, upsampling, 1)
             peak_time = adaptive_centroid(
                 d_waveforms,
-                max(0, min(np.round(peak_index - pz2d).astype(int))),
+                np.clip(
+                    np.round(peak_index - pz2d).astype(int), 0, np.shape(d_waveforms)[1]
```

Comment:
`waveforms.shape[-1]`

Also, clip is inclusive, so it should probably have a -1?

to make the intent clear, it would be more readable to split this into multiple assignments.

Maybe also add a comment why you shift the peak_index by `pz2d`, this is not clear to me. 

You can also avoid a copy by using the `out` arg of `clip`
something like:

```
peak_index = np.round(peak_index - pz2d).astype(int)
n_samples = d_waveforms.shape[-1]
np.clip(peak_index, 0, n_samples - 1, out=peak_index) 
peak_time = adaptive_centroid(d_waveforms, peak_index, leading_edge_rel_descend_limit)
```